velox/common/file/File.cpp

```diff
@@ -238,6 +246,21 @@ uint64_t LocalReadFile::preadv(
   return totalBytesRead;
 }
 
+folly::SemiFuture<uint64_t> LocalReadFile::preadvAsync(
+    uint64_t offset,
+    const std::vector<folly::Range<char*>>& buffers) const {
+  auto promise = std::make_unique<folly::Promise<uint64_t>>();
+  folly::SemiFuture<uint64_t> future = promise->getSemiFuture();
+  executor_->add([this,
+                  _promise = std::move(promise),
+                  _offset = offset,
+                  _buffers = buffers]() {
+    auto delegateFuture = ReadFile::preadvAsync(_offset, _buffers);
+    _promise->setValue(delegateFuture.wait().value());
```

Comment:
nit maybe promise.setTry(delegateFuture.getTry())

to avoid throwing inside the executor

Reply:
Compile failed by
```
auto promise = std::make_unique<folly::Promise<uint64_t>>();
  auto future = promise->getSemiFuture();
  executor_->add([this,
                  _promise = std::move(promise),
                  _offset = offset,
                  _buffers = buffers]() {
    auto delegateFuture = ReadFile::preadvAsync(_offset, _buffers);
    _promise->setTry(delegateFuture.getTry());
  });
```

```
[build] /mnt/DP_disk1/code/velox/velox/common/file/File.cpp: In lambda function:
[build] /mnt/DP_disk1/code/velox/velox/common/file/File.cpp:254:43: error: passing ‘folly::SemiFuture<long unsigned int>’ as ‘this’ argument discards qualifiers [-fpermissive]
[build]   254 |     _promise->setTry(delegateFuture.getTry());
[build]       |                      ~~~~~~~~~~~~~~~~~~~~~^~
[build] In file included from /usr/local/include/folly/futures/Future.h:2648,
[build]                  from /mnt/DP_disk1/code/velox/./velox/common/file/File.h:39,
[build]                  from /mnt/DP_disk1/code/velox/velox/common/file/File.cpp:17:
[build] /usr/local/include/folly/futures/Future-inl.h:2393:8: note:   in call to ‘folly::Try<T> folly::SemiFuture<T>::getTry() && [with T = long unsigned int]’
[build]  2393 | Try<T> SemiFuture<T>::getTry() && {
[build]       |        ^~~~~~~~~~~~~
[build] ninja: build stopped: subcommand failed.
```

Reply:
curious, does `std::move(delegateFuture).getTry()` work?